File: packages/node/shared-infra/src/http/pipeline/step-ssrf-validation.ts
/**
 * @file step-ssrf-validation.ts
 * @description Pipeline Step 3: Rules-Engine Powered SSRF & Target URL Validation.
 */

import type { PipelineStep, PipelineContext } from "./types";
import { validateDestinationUrl } from "../validation/destination-validator";

export class StepSsrfValidation implements PipelineStep {
  public readonly name = "SsrfValidation";
  public readonly description = "Rules-Engine SSRF Destination & IP Allowlist Guard";

  public async execute<T>(ctx: PipelineContext<T>): Promise<void> {
    ctx.stepIndex++;
    await validateDestinationUrl(ctx.config.url, ctx.config.allowedHosts);
  }
}

File: packages/llm-obs-python-packages/python-shared/src/python_shared/http/pipeline/step_ssrf_validation.py
from urllib.parse import urlparse
from python_shared.http.pipeline.types import PipelineContext

class StepSsrfValidation:
    name: str = "SsrfValidation"
    description: str = "SSRF Destination & IP Allowlist Guard"

    async def execute(self, ctx: PipelineContext) -> None:
        ctx.step_index += 1
        parsed = urlparse(ctx.config.url)
        if not parsed.scheme or not parsed.netloc:
            raise ValueError(f"Invalid URL for SSRF validation: {ctx.config.url}")
        if ctx.config.allowed_hosts and parsed.hostname not in ctx.config.allowed_hosts:
            raise ValueError(f"Target host '{parsed.hostname}' is not in allowed hosts list")
